Fails to work on Ubuntu/Debian due to "Duplicate mapping key"

If you run the example on Ubuntu you get:
```
Unhandled exception:
Error on line 25023, column 3: Duplicate mapping key.
      ╷
25023 │   ca_ES:
      │   ^^^^^
      ╵
#0      Loader._loadMapping (package:yaml/src/loader.dart:168:9)
```

This is because appstream-generator was [generating invalid YAML](https://github.com/ximion/appstream-generator/issues/92), specifically the translation merging was leading to duplicate keys in maps (i.e. two ca_ES translations in the above example). While this is now fixed upstream, Ubuntu/Debian releases will continue to have invalid YAML files.

Possible solutions:
- We manage to update all the existing YAML in Ubuntu/Debian - however this can't fix any systems that haven't updated.
- The Dart YAML implementation slackens its requirements on maps (seems extremely unlikely/undesirable).
- We pre-process the YAML to remove these duplicates.
- We make a small YAML decoder in appstream.dart that handles this case.

Running `yamllint` on all the compressed yaml files in `/var/lib/app-info/yaml/` on my machine running kinetic raises 3 duplicated keys in 1 file:

```bash
osomon@dantian:/var/lib/app-info/yaml$ for f in *; do echo $f; zcat -c $f > /tmp/$f; yamllint /tmp/$f | grep duplication; done
es.archive.ubuntu.com_ubuntu_dists_kinetic_main_dep11_Components-amd64.yml.gz
es.archive.ubuntu.com_ubuntu_dists_kinetic_multiverse_dep11_Components-amd64.yml.gz
es.archive.ubuntu.com_ubuntu_dists_kinetic_universe_dep11_Components-amd64.yml.gz
  170229:3  error    duplication of key "oars-1.1" in mapping  (key-duplicates)
  326444:3  error    duplication of key "oars-1.1" in mapping  (key-duplicates)
  401603:3  error    duplication of key "ca_ES" in mapping  (key-duplicates)
es.archive.ubuntu.com_ubuntu_dists_kinetic-updates_main_dep11_Components-amd64.yml.gz
es.archive.ubuntu.com_ubuntu_dists_kinetic-updates_universe_dep11_Components-amd64.yml.gz
security.ubuntu.com_ubuntu_dists_kinetic-security_main_dep11_Components-amd64.yml.gz
security.ubuntu.com_ubuntu_dists_kinetic-security_universe_dep11_Components-amd64.yml.gz
```

I've had a look at the problem, and it seems to me the most sensible approach is option #3 ("We pre-process the YAML to remove these duplicates").

--- a/lib/src/collection.dart
+++ b/lib/src/collection.dart
@@ -378,9 +378,31 @@ class AppstreamCollection {
         components: components);
   }
 
+  // Very dumb removal of invalid YAML documents.
+  // See https://github.com/canonical/appstream.dart/issues/15.
+  // Fixing these documents would be much costlier and error-prone,
+  // hence this simplistic approach to just filter out invalid documents.
+  static String _removeInvalidDocuments(String yaml) {
+    String processNode(String document) {
+      try {
+        loadYamlDocument(document);
+        return document;
+      } on YamlException {
+        return '';
+      }
+    }
+
+    final documentSeparator = '\n---\n';
+    final documents = yaml.split(documentSeparator);
+    for (var i = 0; i < documents.length; ++i) {
+      documents[i] = processNode(documents[i]);
+    }
+    return documents.where((e) => e.isNotEmpty).join(documentSeparator);
+  }
+
   /// Decodes an Appstream collection in YAML format.
   factory AppstreamCollection.fromYaml(String yaml) {
-    var yamlDocuments = loadYamlDocuments(yaml);
+    var yamlDocuments = loadYamlDocuments(_removeInvalidDocuments(yaml));
     if (yamlDocuments.isEmpty) {
       throw FormatException('Empty YAML file');
     }

--- a/test/appstream_test.dart
+++ b/test/appstream_test.dart
@@ -449,6 +449,26 @@ void main() {
         throwsFormatException);
   });
 
+  test('collection - yaml with duplicate mapping keys', () async {
+    var collection = AppstreamCollection.fromYaml("""---
+File: DEP-11
+Version: '0.12'
+Origin: ubuntu-hirsute-main
+---
+Type: console-application
+ID: com.example.Hello
+Package: hello
+Name:
+  C: Hello World
+Summary:
+  C: A simple example application
+ContentRating:
+  oars-1.1: {}
+  oars-1.1: {}
+""");
+    expect(collection.components, isEmpty);
+  });
+
   test('collection - empty - yaml', () async {
     var collection = AppstreamCollection.fromYaml("""---
 File: DEP-11

